dlib/matrix/fft_size.h

```diff
@@ -0,0 +1,215 @@
+        {
+            return _dims.begin() + _size;
+        }
+
+    public:
+        fft_size() = default;
+        /*!
+            ensures
+                - this is properly initialised
+                - size() == 0
+                - is_valid() == false
+        !*/
+        
+        fft_size(std::initializer_list<long> dims)
+        {
+            DLIB_ASSERT(dims.size() > 0, "the initialiser list must be non-empty");
+            DLIB_ASSERT(dims.size() <= _dims.size(), "the initialiser list must have size less than 6");
+            std::copy(dims.begin(), dims.end(), _dims.begin());
+            _size = dims.size();
+            DLIB_ASSERT(is_valid(), "the initialiser list must contain strictly positive values");
+        }
+        /*!
+            requires
+                - dims.size() > 0 and dims.size() <= 5
+                - dims contains strictly positive values
+            ensures
+                - *this is properly initialised
+                - size() == dims.size()
+                - is_valid() == true 
+            throws
+                - dlib::fatal_error if requirements aren't satisfied.
+        !*/
+        
+        size_t size() const
+        {
+            return _size;
+        }
+        /*!
+            ensures
+                - returns the number of elements in *this
+        !*/
+
+        const_reference operator[](size_t index) const
+        {
+            DLIB_ASSERT(index < _size, "index " << index << " out of range [0," << _size << ")");
+            return _dims[index];
+        }
+        /*!
+            requires
+                - size() > 0
+                - index < size()
+            ensures
+                - returns a const reference to the element at position index
+        !*/
+        
+        const_reference back() const
+        {
+            DLIB_ASSERT(_size > 0, "object is empty");
+            return _dims[_size-1];
+        }
+        /*!
+            requires
+                - size() > 0
+            ensures
+                - returns a const reference to (*this)[size()-1]
+        !*/
+        
+        long dimprod() const
+        {
+            DLIB_ASSERT(_size > 0, "object is empty");
+            return std::accumulate(begin(), end(), 1, std::multiplies<long>());
+        }
+        /*!
+            requires
+                - size() > 0
+            ensures
+                - returns the product of all dimensions, i.e. the total number
+                  of elements.
+        !*/
+        
+        bool operator==(const fft_size& other) const
+        {
+            return this->_size == other._size && std::equal(begin(), end(), other.begin());
+        }
+        /*!
+            ensures
+                - returns true if two fft_size objects have same size and same elements.
+        !*/
+        
+        bool is_valid() const
+        {
+            return _size > 0 && std::find_if(begin(), end(), [](long dim) {return dim <= 0;}) == end();
+        }
```

Comment:
This function isn't needed.  The condition can be checked in the constructor.  Then client code can assert `size() != 0`.  Having `is_valid()` and `size()` gives two ways for clients to do the same thing.  But there should be one unambiguous way.

Reply:
Also, negative dims can never happen, because the constructor's preconditions forbid it.  The logic here is making negative dimensions a proper state of this class.  But such `fft_size` objects should never exist at any point.

Reply:
Yep. another good point